src/yarn-commands.ts

```diff
@@ -0,0 +1,71 @@
+import { WriteStream } from 'fs';
+import { debug, runCommand, getStdoutFromCommand } from './misc-utils.js';
+
+/**
+ * Gets the current Yarn version.
+ *
+ * @returns A promise that resolves to the Yarn version string.
+ * @throws An execa error object if the command fails in some way.
+ */
+function getYarnVersion(): Promise<string> {
+  return getStdoutFromCommand('yarn', ['--version']);
+}
+
+/**
+ * Runs `yarn constraints --fix` to autofix all unmet constraints.
+ *
+ * @param repositoryDirectoryPath - The path to the repository directory.
+ * @param stderr - A stream that can be used to write to standard error.
+ * @returns The standard output of the command.
+ * @throws An execa error object if the command fails in some way.
+ */
+export async function fixConstraints(
+  repositoryDirectoryPath: string,
+  stderr: Pick<WriteStream, 'write'>,
+): Promise<void> {
+  const version = await getYarnVersion();
+  const majorVersion = parseInt(version.split('.')[0], 10);
+
+  if (majorVersion >= 2) {
```

Comment:
I think we can safely assume that the Yarn version is >= 2. We don't use this tool for any other projects at the moment. So perhaps we can drop this check? (The `--no-immutable` option to `yarn` and the `yarn dedupe` commands already assume Yarn >= 2 anyway.)